packages/lib/services/e2ee/crypto.ts

```diff
@@ -0,0 +1,132 @@
+import { _ } from '../../locale';
+import { Crypto, CryptoBuffer, Digest, CipherAlgorithm, EncryptionResult } from './types';
+import { promisify } from 'util';
+import {
+	getCiphers as nodeGetCiphers, getHashes as nodeGetHashes,
+	randomBytes as nodeRandomBytes,
+	pbkdf2 as nodePbkdf2,
+	createCipheriv, createDecipheriv,
+	CipherGCMOptions, CipherGCM, DecipherGCM,
+} from 'crypto';
+
+const crypto: Crypto = {
+
+	getCiphers: () => {
+		return nodeGetCiphers();
+	},
+
+	getHashes: () => {
+		return nodeGetHashes();
+	},
+
+	randomBytes: async (size: number) => {
+		const randomBytesAsync = promisify(nodeRandomBytes);
+		return randomBytesAsync(size);
+	},
+
+	pbkdf2Raw: async (password: string, salt: CryptoBuffer, iterations: number, keylen: number, digest: Digest) => {
+		const pbkdf2Async = promisify(nodePbkdf2);
+		return pbkdf2Async(password, salt, iterations, keylen, digest);
+	},
+
+
+	encryptRaw: (data: CryptoBuffer, algorithm: CipherAlgorithm, key: CryptoBuffer, iv: CryptoBuffer | null, authTagLength: number, associatedData: CryptoBuffer | null) => {
+
+		if (associatedData === null) {
+			associatedData = Buffer.alloc(0);
+		}
+
+		let cipher = null;
+		if (algorithm === CipherAlgorithm.AES_256_GCM || algorithm === CipherAlgorithm.AES_192_GCM || algorithm === CipherAlgorithm.AES_128_GCM) {
+			cipher = createCipheriv(algorithm, key, iv, { authTagLength: authTagLength } as CipherGCMOptions) as CipherGCM;
+			iv = iv || nodeRandomBytes(12); // "For IVs, it is recommended that implementations restrict support to the length of 96 bits, to promote interoperability, efficiency, and simplicity of design." - NIST SP 800-38D
+		} else {
+			throw new Error(_('Unknown cipher algorithm: %s', algorithm));
```

Comment:
Why is this check needed? Isn't TypeScript sufficient to ensure that the input is correct?

Reply:
I removed some unused algorithms/branches here but I didn't remove the if statements there. I've cleaned it up in commit bca6e4da94e8736d21b09b6ecd954ee2a7b8d613.